File: src/cli/commands/context.ts
import { Command } from 'commander'

import { getDb } from '../../db/index.js'
import { createCustomFieldRepo } from '../../db/repositories/custom-field.repo.js'
import { createDealRepo } from '../../db/repositories/deal.repo.js'
import { createInteractionRepo } from '../../db/repositories/interaction.repo.js'
import { createOrgRepo } from '../../db/repositories/org.repo.js'
import { createPersonRepo } from '../../db/repositories/person.repo.js'
import { createRelationshipRepo } from '../../db/repositories/relationship.repo.js'
import { createTagRepo } from '../../db/repositories/tag.repo.js'
import { createTaskRepo } from '../../db/repositories/task.repo.js'
import { resolveFormat } from '../../formatters/index.js'
import { NotFoundError } from '../../models/errors.js'

interface GlobalOpts {
  format?: string
  quiet?: boolean
  db?: string
  color?: boolean
}

function getOpts(cmd: Command): GlobalOpts {
  let root = cmd
  while (root.parent) root = root.parent
  return root.opts()
}

export function registerContextCommands(program: Command): void {
  program
    .command('context')
    .description('Full context briefing for a person')
    .argument('<id>', 'Person ID')
    .option('--limit <n>', 'Max interactions to show', '10')
    .action((idStr: string, opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const id = Number(idStr)
      const limit = Number(opts['limit'])

      const personRepo = createPersonRepo(db)
      const orgRepo = createOrgRepo(db)
      const interactionRepo = createInteractionRepo(db)
      const dealRepo = createDealRepo(db)
      const taskRepo = createTaskRepo(db)
      const tagRepo = createTagRepo(db)
      const relRepo = createRelationshipRepo(db)
      const cfRepo = createCustomFieldRepo(db)

      const person = personRepo.findById(id)
      if (!person) throw new NotFoundError('person', id)

      const org = person.org_id ? orgRepo.findById(person.org_id) : undefined
      const interactions = interactionRepo.findAll({ personId: id, limit })
      const deals = dealRepo.findAll({ personId: id })
      const tasks = taskRepo.findAll({ personId: id, completed: false })
      const tags = tagRepo.getForEntity('person', id)
      const relationships = relRepo.findForPerson(id)
      const customFields = cfRepo.get('person', id)

      const format = resolveFormat(globalOpts.format)

      if (format === 'json') {
        const data = {
          person,
          organization: org ?? null,
          tags: tags.map((t) => t.name),
          custom_fields: customFields,
          relationships: relationships.map((r) => ({
            id: r.id,
            person_id: r.person_id === id ? r.related_person_id : r.person_id,
            type: r.type,
            notes: r.notes,
          })),
          recent_interactions: interactions.map((i) => ({ ...i })),
          open_deals: deals.filter((d) => d.stage !== 'won' && d.stage !== 'lost'),
          open_tasks: tasks,
          stats: {
            total_interactions: interactions.length,
            total_deals: deals.length,
            open_tasks: tasks.length,
          },
        }
        process.stdout.write(JSON.stringify(data, null, process.stdout.isTTY ? 2 : 0) + '\n')
      } else {
        const name = [person.first_name, person.last_name].filter(Boolean).join(' ')
        const lines: string[] = []

        lines.push(`=== ${name} ===`)
        if (person.email) lines.push(`Email: ${person.email}`)
        if (person.phone) lines.push(`Phone: ${person.phone}`)
        if (person.title) lines.push(`Title: ${person.title}`)
        if (person.company) lines.push(`Company: ${person.company}`)
        if (org) lines.push(`Organization: ${org.name}`)
        if (person.location) lines.push(`Location: ${person.location}`)
        if (tags.length > 0) lines.push(`Tags: ${tags.map((t) => t.name).join(', ')}`)
        if (person.summary) lines.push(`\nSummary: ${person.summary}`)
        if (person.notes) lines.push(`Notes: ${person.notes}`)

        if (customFields.length > 0) {
          lines.push('\n--- Custom Fields ---')
          for (const cf of customFields) {
            lines.push(`${cf.field_name}: ${cf.field_value ?? ''}`)
          }
        }

        if (relationships.length > 0) {
          lines.push('\n--- Relationships ---')
          for (const r of relationships) {
            const otherId = r.person_id === id ? r.related_person_id : r.person_id
            const other = personRepo.findById(otherId)
            const otherName = other
              ? [other.first_name, other.last_name].filter(Boolean).join(' ')
              : `#${String(otherId)}`
            lines.push(`${r.type}: ${otherName}`)
          }
        }

        if (interactions.length > 0) {
          lines.push('\n--- Recent Interactions ---')
          for (const i of interactions) {
            const subject = i.subject ?? '(no subject)'
            lines.push(`${i.occurred_at} [${i.type}] ${subject}`)
          }
        }

        const openDeals = deals.filter((d) => d.stage !== 'won' && d.stage !== 'lost')
        if (openDeals.length > 0) {
          lines.push('\n--- Open Deals ---')
          for (const d of openDeals) {
            const value = d.value ? ` ($${String(d.value)})` : ''
            lines.push(`#${String(d.id)} ${d.title} [${d.stage}]${value}`)
          }
        }

        if (tasks.length > 0) {
          lines.push('\n--- Open Tasks ---')
          for (const t of tasks) {
            const due = t.due_at ? ` due ${t.due_at}` : ''
            lines.push(`#${String(t.id)} ${t.title} [${t.priority}]${due}`)
          }
        }

        process.stdout.write(lines.join('\n') + '\n')
      }
    })
}

File: internal/cli/context.go
package cli

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/jdanielnd/crm-cli/internal/db/repo"
	"github.com/jdanielnd/crm-cli/internal/format"
	"github.com/jdanielnd/crm-cli/internal/model"
	"github.com/spf13/cobra"
)

func registerContextCommand(rootCmd *cobra.Command) {
	cmd := &cobra.Command{
		Use:   "context <person_id_or_name>",
		Short: "Full context briefing for a person",
		Long:  "Returns person profile, organization, recent interactions, open deals, open tasks, relationships, and tags.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			pr := repo.NewPersonRepo(db)

			// Try to parse as ID first, otherwise search by name
			var person *model.Person
			id, parseErr := parseEntityID(args[0], "person")
			if parseErr == nil {
				person, err = pr.FindByID(cmd.Context(), id)
				if err != nil {
					return err
				}
			} else {
				people, err := pr.Search(cmd.Context(), args[0], 1)
				if err != nil {
					return err
				}
				if len(people) == 0 {
					return model.NewExitError(model.ErrNotFound, "no person found matching %q", args[0])
				}
				person = people[0]
			}

			ctx := cmd.Context()
			result := map[string]any{}

			// Person profile
			result["person"] = personToMap(person)

			// Organization
			var org *model.Organization
			if person.OrgID != nil {
				or := repo.NewOrgRepo(db)
				org, err = or.FindByID(ctx, *person.OrgID)
				if err != nil {
					return err
				}
				result["organization"] = orgToMap(org)
			}

			// Recent interactions
			ir := repo.NewInteractionRepo(db)
			interactions, err := ir.FindAll(ctx, model.InteractionFilters{PersonID: &person.ID, Limit: 10})
			if err != nil {
				return err
			}
			if len(interactions) > 0 {
				result["recent_interactions"] = interactionsToMaps(interactions)
			}

			// Open deals
			dr := repo.NewDealRepo(db)
			deals, err := dr.FindAll(ctx, model.DealFilters{PersonID: &person.ID})
			if err != nil {
				return err
			}
			if len(deals) > 0 {
				result["deals"] = dealsToMaps(deals)
			}

			// Open tasks
			tr := repo.NewTaskRepo(db)
			tasks, err := tr.FindAll(ctx, model.TaskFilters{PersonID: &person.ID})
			if err != nil {
				return err
			}
			if len(tasks) > 0 {
				result["tasks"] = tasksToMaps(tasks)
			}

			// Relationships
			rr := repo.NewRelationshipRepo(db)
			rels, err := rr.FindForPerson(ctx, person.ID)
			if err != nil {
				return err
			}
			if len(rels) > 0 {
				relMaps := make([]map[string]any, len(rels))
				for i, rel := range rels {
					relMaps[i] = map[string]any{
						"id":                rel.ID,
						"person_id":         rel.PersonID,
						"related_person_id": rel.RelatedPersonID,
						"type":              rel.Type,
					}
					if rel.Notes != nil {
						relMaps[i]["notes"] = *rel.Notes
					}
				}
				result["relationships"] = relMaps
			}

			// Tags
			tagRepo := repo.NewTagRepo(db)
			tags, err := tagRepo.GetForEntity(ctx, "person", person.ID)
			if err != nil {
				return err
			}
			if len(tags) > 0 {
				tagNames := make([]string, len(tags))
				for i, t := range tags {
					tagNames[i] = t.Name
				}
				result["tags"] = tagNames
			}

			// JSON/CSV/TSV: output as structured data
			f := resolveFormat()
			if f == format.FormatJSON || f == format.FormatCSV || f == format.FormatTSV {
				enc := json.NewEncoder(os.Stdout)
				enc.SetIndent("", "  ")
				return enc.Encode(result)
			}

			// Table/human-readable: custom formatted output
			w := os.Stdout
			name := person.FirstName
			if person.LastName != nil {
				name += " " + *person.LastName
			}
			fmt.Fprintf(w, "# %s (ID: %d)\n", name, person.ID)

			if person.Email != nil {
				fmt.Fprintf(w, "  Email: %s\n", *person.Email)
			}
			if person.Phone != nil {
				fmt.Fprintf(w, "  Phone: %s\n", *person.Phone)
			}
			if person.Title != nil {
				fmt.Fprintf(w, "  Title: %s\n", *person.Title)
			}
			if person.Company != nil {
				fmt.Fprintf(w, "  Company: %s\n", *person.Company)
			}
			if person.Location != nil {
				fmt.Fprintf(w, "  Location: %s\n", *person.Location)
			}
			if org != nil {
				fmt.Fprintf(w, "  Organization: %s (#%d)\n", org.Name, org.ID)
			}
			if person.Summary != nil {
				fmt.Fprintf(w, "  Summary: %s\n", *person.Summary)
			}

			if tagNames, ok := result["tags"].([]string); ok && len(tagNames) > 0 {
				fmt.Fprintf(w, "\nTags: %s\n", strings.Join(tagNames, ", "))
			}

			if len(interactions) > 0 {
				fmt.Fprintf(w, "\nRecent Interactions (%d):\n", len(interactions))
				for _, i := range interactions {
					subject := ""
					if i.Subject != nil {
						subject = *i.Subject
					}
					dir := ""
					if i.Direction != nil {
						dir = " [" + *i.Direction + "]"
					}
					fmt.Fprintf(w, "  #%-4d %s: %s%s (%s)\n", i.ID, i.Type, subject, dir, i.OccurredAt)
				}
			}

			if len(deals) > 0 {
				fmt.Fprintf(w, "\nDeals (%d):\n", len(deals))
				for _, d := range deals {
					value := ""
					if d.Value != nil {
						value = fmt.Sprintf(" $%.0f", *d.Value)
					}
					fmt.Fprintf(w, "  #%-4d %s [%s]%s\n", d.ID, d.Title, d.Stage, value)
				}
			}

			if len(tasks) > 0 {
				fmt.Fprintf(w, "\nOpen Tasks (%d):\n", len(tasks))
				for _, t := range tasks {
					due := ""
					if t.DueAt != nil {
						due = " (due " + *t.DueAt + ")"
					}
					fmt.Fprintf(w, "  #%-4d %s [%s]%s\n", t.ID, t.Title, t.Priority, due)
				}
			}

			if len(rels) > 0 {
				fmt.Fprintf(w, "\nRelationships (%d):\n", len(rels))
				for _, rel := range rels {
					otherID := rel.RelatedPersonID
					if otherID == person.ID {
						otherID = rel.PersonID
					}
					note := ""
					if rel.Notes != nil {
						note = " — " + *rel.Notes
					}
					fmt.Fprintf(w, "  Person #%d (%s)%s\n", otherID, rel.Type, note)
				}
			}

			return nil
		},
	}

	rootCmd.AddCommand(cmd)
}

